apps/server/src/controllers/integrationController.ts

```diff
@@ -48,62 +47,37 @@ const actionHandlers: Record<string, ActionHandler> = {
-    }
-    const newState = messageService.setTimerMessage(payload);
-    return { payload: newState.timerMessage };
-  },
-  'set-timer-blink': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setTimerBlink(Boolean(payload));
-    return { payload: newState.timerMessage };
-  },
-  'set-timer-blackout': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setTimerBlackout(Boolean(payload));
-    return { payload: newState.timerMessage };
-  },
-  'set-timer-message-text': (payload) => {
-    assert.isString(payload);
-    const newState = messageService.setTimerText(payload);
-    return { payload: newState.timerMessage };
-  },
-  'set-timer-message-visible': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setTimerVisibility(Boolean(payload));
-    return { payload: newState.timerMessage };
-  },
-  'set-public-message-text': (payload) => {
-    assert.isString(payload);
-    const newState = messageService.setPublicText(payload);
-    return { payload: newState.publicMessage };
-  },
-  'set-public-message-visible': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setPublicVisibility(Boolean(payload));
-    return { payload: newState.publicMessage };
-  },
-  'set-lower-message-text': (payload) => {
-    assert.isString(payload);
-    const newState = messageService.setLowerText(payload);
-    return { payload: newState.lowerMessage };
-  },
-  'set-lower-message-visible': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setLowerVisibility(Boolean(payload));
-    return { payload: newState.lowerMessage };
-  },
-  'set-external-message-text': (payload) => {
-    assert.isString(payload);
-    const newState = messageService.setExternalText(payload);
-    return { payload: newState.externalMessage };
-  },
-  'set-external-message-visible': (payload) => {
-    assert.isDefined(payload);
-    const newState = messageService.setExternalVisibility(Boolean(payload));
-    return { payload: newState.externalMessage };
+  message: (payload) => {
+    const reply = { payload: {} };
+    Object.keys(payload).forEach((key) => {
+      if (!isPartialTimerMessage(payload[key])) {
+        throw new Error('Message: Payload is not valid');
+      }
+      let newState;
+      switch (key) {
+        case 'timer': {
+          newState = messageService.setTimerMessage(payload[key]);
+          break;
+        }
+        case 'public': {
+          newState = messageService.setPublicMessage(payload[key]);
+          break;
+        }
+        case 'external': {
+          newState = messageService.setExternalMessage(payload[key]);
+          break;
+        }
+        case 'lower': {
+          newState = messageService.setLowerMessage(payload[key]);
+          break;
+        }
+        default: {
+          throw new Error(`Message: ${key} dose not exist`);
```

Comment:
If I understand correctly, any bad property would fail the rest of the request, but changes before would have affected. so

```ts
{
  visible: true,
  text: "sadasdas",
  not-value: "going to throw"
}
```
would both throw an error and make changes

I also not sure it is a good idea to affect changes one by one, since this starts a sequence of async actions to notify the clients.
How about composing a patch object. We would then make a single call to validate the object as a whole and another call to make a single change to the message service

Also: do we really need to reuse the code here?
I think separating this into three different `timer` / `public` / `external` / `lower` handlers is fine. The APIs are sometimes different and when they are the same it is mostly coincidental

Reply:
> composing a patch object
I was looking into that, but we have the `throttle` function on the message service
I thing the different services should be throttled separately? maybe it is still possible?

> Also: do we really need to reuse the code here?

I might be thinking to far ahead but maybe the message service would also end up being user definable?

Reply:
> I was looking into that, but we have the throttle function on the message service
> I thing the different services should be throttled separately? maybe it is still possible?


I dont think this stops us from creating the single patch object, I think we should do it. So instead of iterating through the keys, we make a single object (without loops) and a single patch request.
We dont need loops since the keys are not dynamic and we know which ones they are, this is also a good way of filtering out bad keys

> I might be thinking to far ahead but maybe the message service would also end up being user definable?

Maybe, but we should worry about that later. How would that work?

Reply:
> how do you see that could work

I could imagine a backstage view in the Green room and in the tech break room
or maybe both the a speaker and the moderators have a screen each with a timer view
if there then were a parameter for the views where you select the message source (kind of like the new lower source)
then you can send separate messages to each